File: Python3/601 - 800/606 Construct String from Binary Tree.py
# 68ms, 92.27%
class Solution:
    def __init__(self):
        self.res = []

    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """

        def preorder(node):
            if not node: return
            self.res.append(str(node.val))
            if node.left or node.right:
                self.res.append('(')
                preorder(node.left)
                self.res.append(')')
            if node.right:
                self.res.append('(')
                preorder(node.right)
                self.res.append(')')

        preorder(t)
        return ''.join(self.res)
        

# 72ms, 63.64%
class Solution:
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if not t: return ''
        left = '({})'.format(self.tree2str(t.left)) if t.left or t.right else ''
        right = '({})'.format(self.tree2str(t.right)) if t.right else ''
        return '{}{}{}'.format(t.val, left, right)
    

File: 601 - 800/606 Construct String from Binary Tree.cc
// 17ms, 85.06%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

static int x = []() {
    ios_base::sync_with_stdio(false); // toggles off the synchronization
    cin.tie(nullptr);                 // ties cin with nullptr, not cout
    return 0;
}();

class Solution {
private:
    string res = "";

public:
    string tree2str(TreeNode *t) {
        if (!t) return "";
        helper(t);
        return res;
    }

private:
    void helper(TreeNode *t) {
        res += to_string(t->val);
        if (t->left || t->right) {
            res += "(";
            tree2str(t->left);
            res += ")";
            if (t->right) {
                res += "(";
                tree2str(t->right);
                res += ")";
            }
        }
    }
};
